src/openzeppelin/token/erc20.cairo

```diff
@@ -0,0 +1,200 @@
+        ERC20::approve(spender, amount)
+    }
+
+    #[external]
+    fn increase_allowance(spender: ContractAddress, added_value: u256) -> bool {
+        _increase_allowance(spender, added_value)
+    }
+
+    #[external]
+    fn decrease_allowance(spender: ContractAddress, subtracted_value: u256) -> bool {
+        _decrease_allowance(spender, subtracted_value)
+    }
+
+    ///
+    /// Internals
+    ///
+
+    fn initializer(name_: felt, symbol_: felt, initial_supply: u256, recipient: ContractAddress) {
+        _name::write(name_);
+        _symbol::write(symbol_);
+        _mint(recipient, initial_supply);
+    }
+
+    fn _increase_allowance(spender: ContractAddress, added_value: u256) -> bool {
+        let caller = get_caller_address();
+        _approve(caller, spender, _allowances::read((caller, spender)) + added_value);
+        true
+    }
+
+    fn _decrease_allowance(spender: ContractAddress, subtracted_value: u256) -> bool {
+        let caller = get_caller_address();
+        _approve(caller, spender, _allowances::read((caller, spender)) - subtracted_value);
+        true
+    }
+
+    fn _mint(recipient: ContractAddress, amount: u256) {
+        assert(!recipient.is_zero(), 'ERC20: mint to 0');
+        _total_supply::write(_total_supply::read() + amount);
+        _balances::write(recipient, _balances::read(recipient) + amount);
+        Transfer(contract_address_const::<0>(), recipient, amount);
+    }
+
+    fn _burn(account: ContractAddress, amount: u256) {
+        assert(!account.is_zero(), 'ERC20: burn from 0');
+        _total_supply::write(_total_supply::read() - amount);
+        _balances::write(account, _balances::read(account) - amount);
+        Transfer(account, contract_address_const::<0>(), amount);
+    }
+
+    fn _approve(owner: ContractAddress, spender: ContractAddress, amount: u256) {
+        assert(!owner.is_zero(), 'ERC20: approve from 0');
+        assert(!spender.is_zero(), 'ERC20: approve to 0');
+        _allowances::write((owner, spender), amount);
+        Approval(owner, spender, amount);
+    }
+
+    fn _transfer(sender: ContractAddress, recipient: ContractAddress, amount: u256) {
+        assert(!sender.is_zero(), 'ERC20: transfer from 0');
+        assert(!recipient.is_zero(), 'ERC20: transfer to 0');
+        _balances::write(sender, _balances::read(sender) - amount);
+        _balances::write(recipient, _balances::read(recipient) + amount);
+        Transfer(sender, recipient, amount);
+    }
+
+    fn _spend_allowance(owner: ContractAddress, spender: ContractAddress, amount: u256) {
+        let current_allowance = _allowances::read((owner, spender));
+        let ONES_MASK = 0xffffffffffffffffffffffffffffffff_u128;
+        let is_unlimited_allowance =
+            current_allowance.low == ONES_MASK & current_allowance.high == ONES_MASK;
+        if !is_unlimited_allowance {
+            _approve(owner, spender, current_allowance - amount);
+        }
+    }
```

Comment:
With this proposal https://github.com/starkware-libs/cairo/pull/2563 we could do:

```suggestion
    fn _spend_allowance(owner: ContractAddress, spender: ContractAddress, amount: u256) {
        let current_allowance = _allowances::read((owner, spender));
        if is_unlimited_allowance != max_uint256() {
            _approve(owner, spender, current_allowance - amount);
        }
    }
```